internal/runners/run/run.go

```diff
@@ -113,16 +99,31 @@ func run(out output.Outputer, subs subshell.SubShell, name string, args []string
 		}
 		subs.SetEnv(env)
 
-		// get the "clean" path (only PATHS that are set by venv)
+		// search the "clean" path first (PATHS that are set by venv)
 		env, err = venv.GetEnv(false, "")
 		if err != nil {
 			return err
 		}
-		path = env["PATH"]
+		path = env["PATH"] + string(os.PathSeparator) + path
 	}
 
-	if !langExec.Builtin() && !pathProvidesExec(configCachePath(), langExec.Name(), path) {
-		return FailExecNotFound.New("error_state_run_unknown_exec")
+	for _, l := range script.Languages() {
+		if pathProvidesLang(path, l) {
+			lang = l
+			break
+		}
+		attempted = append(attempted, l.String())
+	}
+
+	if !lang.Recognized() {
+		warning := locale.Tl(
+			"run_warn_deprecated_script_without_language",
+			"[YELLOW]DEPRECATION WARNING: Scripts without a recognized or defined language currently fall back to using the default shell for your platform. This fallback mechanic will soon stop working and a language will need to be explicitly defined for each script. Please configure the 'language' field with a valid option (one of {{.V0}})[/RESET]",
```

Comment:
You're handling two different conditions under the same warning.

Behaviour 1 = no language defined, print warning and default to something.

Behaviour 2 = language not valid, return error and exit nonzero.